lib/THC/THCReduceApplyUtils.cuh

```diff
@@ -72,6 +72,76 @@ __device__ T reduceBlock(T* smem,
   return r;
 }
 
+// Return value is in threadVals for threadIdx.x == 0
+template <typename T, typename ReduceOp, int N>
+__device__ void reduceNBlock(T *smem,
+                             T threadVals[N],
+                             int numVals,
+                             ReduceOp reduceOp,
+                             T init) {
+  if (numVals == 0) {
+#pragma unroll
+    for (int i = 0; i < N; ++i) {
+      threadVals[i] = init;
+    }
+    return;
+  }
+
+  if (threadIdx.x < numVals) {
+#pragma unroll
+    for (int i = 0; i < N; ++i) {
+      smem[threadIdx.x * N + i] = threadVals[i];
```

Comment:
I think this would be slightly faster if the threads were writing contiguously:

`smem[i * numVals + threadIdx.x] = threadVals[i]`